### DISCUSSION TITLE
Release 1.2.5

### DISCUSSION THREAD

**Discussion by ThomasWaldmann** [borgbackup/borg#7787] [General]
## borgbackup 1.2.5 release

This release includes a security fix plus related upgrade instructions at the top of the change log!

Long changelog:

https://github.com/borgbackup/borg/blob/1.2.5/docs/changes.rst#version-125-2023-08-30

Short borg 1.2 overview (from a borg 1.1 perspective):

https://www.borgbackup.org/releases/borg-1.2.html

### Installation

If you use pip to install this, use: pip install pkgconfig ; pip install "borgbackup==1.2.5"

For other installation methods and more details, please see: https://borgbackup.org/

### Feedback

<hr /><em>This discussion was created from the release <a href='https://github.com/borgbackup/borg/releases/tag/'>Release 1.2.5</a>.</em>

**Comment by ipha** [comment]
I'm getting `Data integrity error: Archive authentication did not verify` when I run `borg list --format='{name} {time} tam:{tam}{NL}' <repository>`.

I'll revert to 1.2.4, run a full check and try again.

check on 1.2.4 returned errors consistent with issue https://github.com/borgbackup/borg/issues/6687. Repair fixed them, but 1.2.5 still returns `Data integrity error: Archive authentication did not verify`

**Reply by ThomasWaldmann** [reply]
That's strange. I practically tried the 1.2.5 code and the upgrade steps with all sorts of encrypted/not-encrypted repos of borg versions 1.0, 1.1 and 1.2.

**Reply by ThomasWaldmann** [reply]
What package / binary of 1.2.5 did you try when you got that error msg? Do you have a full traceback?

**Reply by ThomasWaldmann** [reply]
And did you follow the upgrade steps (see top of changelog) exactly (esp. the first 2)?

**Reply by ThomasWaldmann** [reply]
Tried to reproduce. I used borg 1.0.8 to create a repo and an archive (that borg is old enough to not create TAM authentication tag for archives). Then I used the 1.2.5 release fat binary for macOS to follow the upgrade steps:

```
% export BORG_REPO=/tmp/borg10
% export BORG_PASSPHRASE=1                                                                 

% ./borg-1.0.8 init -e repokey
% ./borg-1.0.8 create ::archive_no_tam README.rst 
% ./borg-1.0.8 list                        
archive_no_tam                       Wed, 2023-08-30 18:53:33

% dist/borg-macos64 -V
borg-macos64 1.2.5

% dist/borg-macos64 info --debug 2>&1 | grep TAM | grep -i manifest 
Manifest TAM not found and not required

% dist/borg-macos64 upgrade --tam --force                     
Manifest contents:
archive_no_tam                       Wed, 2023-08-30 18:53:33 [e7ceb5e1335348530bd9f56e9df4d4e3f33356f9054997a326a189d8ab86cfa8]
Key updated
Key location: /tmp/borg10
Updated security database

% dist/borg-macos64 list --format='{name} {time} tam:{tam}{NL}' 
archive_no_tam Wed, 2023-08-30 18:53:33 tam:none

% dist/borg-macos64 upgrade --archives-tam                
Added archive TAM:   archive_no_tam                       Wed, 2023-08-30 18:53:33 [e7ceb5e1335348530bd9f56e9df4d4e3f33356f9054997a326a189d8ab86cfa8] -> [39d2c46b30d08a315853dd2b779478ab07c07c13e9dc02f261ae27735d69d4a4]

% dist/borg-macos64 list --format='{name} {time} tam:{tam}{NL}'
archive_no_tam Wed, 2023-08-30 18:53:33 tam:verified

% dist/borg-macos64 check --repair
This is a potentially dangerous function.
check --repair might lead to data loss (for kinds of corruption it is not
capable of dealing with). BE VERY CAREFUL!

Type 'YES' if you understand this and want to continue: YES

% dist/borg-macos64 list                                       
archive_no_tam                       Wed, 2023-08-30 18:53:33 [39d2c46b30d08a315853dd2b779478ab07c07c13e9dc02f261ae27735d69d4a4]

% dist/borg-macos64 list ::archive_no_tam
-rw-r--r-- tw     staff      8119 Wed, 2023-08-30 18:35:26 README.rst
```

So, nothing special, worked for me as expected.

**Reply by skyegecko** [reply]
I am encountering the same error on 1.2.5. After rolling back to 1.2.4 and performing a `check` I found no errors at all. A little traceback, maybe this will mean something to you:

```
geckobackup@alchemist ~ [2]> borg info --debug repos/windowscygwin/
using builtin fallback logging configuration
33 self tests completed in 0.07 seconds
Verified integrity of /media/raid/geckobackup/repos/windowscygwin/index.217711
Enter passphrase for key /media/raid/geckobackup/repos/windowscygwin:
TAM-verified manifest
security: read previous location '/media/raid/geckobackup/repos/windowscygwin'
security: read manifest timestamp '2023-08-30T17:34:50.167148'
security: determined newest manifest timestamp as 2023-08-30T17:34:50.167148
security: repository checks ok, allowing access
Verified integrity of /media/raid/geckobackup/.cache/borg/XXXXX
security: read previous location '/media/raid/geckobackup/repos/windowscygwin'
security: read manifest timestamp '2023-08-30T17:34:50.167148'
security: determined newest manifest timestamp as 2023-08-30T17:34:50.167148
security: repository checks ok, allowing access
Repository ID: XXXXX
Location: /media/raid/geckobackup/repos/windowscygwin
Encrypted: Yes (repokey)
Cache: /media/raid/geckobackup/.cache/borg/XXXXX
Security dir: /media/raid/geckobackup/.config/borg/security/XXXXX
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
Data integrity error: Archive authentication did not verify
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/borg/archiver.py", line 5324, in main
    exit_code = archiver.run(args)
  File "/usr/local/lib/python3.9/dist-packages/borg/archiver.py", line 5255, in run
    return set_ec(func(args))
  File "/usr/local/lib/python3.9/dist-packages/borg/archiver.py", line 181, in wrapper
    return method(self, args, repository=repository, cache=cache_, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/borg/archiver.py", line 1440, in do_info
    return self._info_repository(args, repository, manifest, key, cache)
  File "/usr/local/lib/python3.9/dist-packages/borg/archiver.py", line 1520, in _info_repository
    print(str(cache))
  File "/usr/local/lib/python3.9/dist-packages/borg/cache.py", line 418, in __str__
    return self.str_format.format(self.format_tuple())
  File "/usr/local/lib/python3.9/dist-packages/borg/cache.py", line 445, in format_tuple
    stats = self.stats()
  File "/usr/local/lib/python3.9/dist-packages/borg/cache.py", line 435, in stats
    archive = Archive(self.repository, self.key, self.manifest, archive_name,
  File "/usr/local/lib/python3.9/dist-packages/borg/archive.py", line 488, in __init__
    self.load(info.id)
  File "/usr/local/lib/python3.9/dist-packages/borg/archive.py", line 501, in load
    self.metadata = self._load_meta(self.id)
  File "/usr/local/lib/python3.9/dist-packages/borg/archive.py", line 493, in _load_meta
    archive, self.tam_verified, _ = self.key.unpack_and_verify_archive(data, force_tam_not_required=True)
  File "/usr/local/lib/python3.9/dist-packages/borg/crypto/key.py", line 325, in unpack_and_verify_archive
    raise ArchiveTAMInvalid()
borg.crypto.key.ArchiveTAMInvalid: Data integrity error: Archive authentication did not verify

Platform: Linux alchemist 5.10.0-19-amd64 #1 SMP Debian 5.10.149-2 (2022-10-21) x86_64
Linux: Unknown Linux
Borg: 1.2.5  Python: CPython 3.9.2 msgpack: 1.0.4 fuse: None [pyfuse3,llfuse]
PID: 3892234  CWD: /media/raid/geckobackup
sys.argv: ['/usr/local/bin/borg', 'info', '--debug', 'repos/windowscygwin/']
SSH_ORIGINAL_COMMAND: None
```

**Reply by ThomasWaldmann** [reply]
Interesting. You seem to have an archive that **is TAM authenticated**, but the 1.2.5 code can't verify the authentication (the computed hmac differs from the encountered hmac).

**Reply by ThomasWaldmann** [reply]
Can you tell something about how these archives were created? Anything special? Which borg versions did you use?

**Reply by skyegecko** [reply]
The repository has been around for a while, since 1.10 if memory serves me correctly. To my knowledge I've always carried out the additional upgrade steps when they occurred. Today I upgraded directly from 1.2.3 to 1.2.5.

A typical daily create looks like this:

    Executing: borg create --verbose --stats --show-rc --show-version --compression auto,zstd,5 --exclude-from exclude_users --exclude-caches --upload-buffer 512 --noflags ::users_20230830_2020 /c/Users/gecko

**Reply by ThomasWaldmann** [reply]
Good and bad news:

Good: could reproduce the issue.

`borg rename` did not recompute a TAM authentication tag for the resulting archive, but just kept the (now invalid) previous one.  `borg recreate` is internally also using rename, so causing the same issue.

@dutchgecko @ipha guess you maybe have used `borg recreate` or `borg rename` in the problematic repo?

Bad: guess this will need a 1.2.6 release to fix.

**Reply by ThomasWaldmann** [reply]
Filed issue #7791 about this.

Everybody encountering `borg.crypto.key.ArchiveTAMInvalid: Data integrity error: Archive authentication did not verify` while upgrading to borg 1.2.5 must abort the upgrade and revert to borg <= 1.2.4 until a fix is released!

**Reply by ipha** [reply]
Yes, rename is called monthly as part of my backup script.

**Reply by skyegecko** [reply]
I have indeed done a `borg recreate` in the past to change compression type.

**Reply by ThomasWaldmann** [reply]
borg 1.2.6 with a fix was just released!
